File: agentsniff/notifier.py
"""
AgentSniff - Alert notifications (webhook + SMTP email).

Sends alerts when scans detect agents matching the configured thresholds.
Uses aiohttp (already a project dependency) for webhooks and stdlib smtplib
for email (via run_in_executor to avoid blocking the event loop).
"""

from __future__ import annotations

import asyncio
import json
import logging
import smtplib
import time
from datetime import datetime, timezone
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from typing import Any

import aiohttp

from agentsniff.config import ScanConfig

logger = logging.getLogger("agentsniff.notifier")

# Confidence levels ordered for threshold comparison
_CONFIDENCE_ORDER = {"low": 0, "medium": 1, "high": 2, "confirmed": 3}

# Track last alert time for cooldown
_last_alert_time: float = 0.0


def should_alert(result: Any, config: ScanConfig) -> bool:
    """Check whether a scan result should trigger alerts."""
    global _last_alert_time

    if not config.alert_enabled:
        return False

    # Must have at least one alert channel configured
    if not config.webhook_url and not config.smtp_to:
        return False

    summary = result.summary
    total = summary.get("total_agents", 0)

    if total < config.alert_min_agents:
        return False

    # Check if any agent meets the minimum confidence threshold
    min_level = _CONFIDENCE_ORDER.get(config.alert_min_confidence, 0)
    by_conf = summary.get("by_confidence", {})
    qualifying = sum(
        count for level, count in by_conf.items()
        if _CONFIDENCE_ORDER.get(level, 0) >= min_level
    )
    if qualifying < config.alert_min_agents:
        return False

    # Cooldown check
    if config.alert_cooldown > 0:
        now = time.monotonic()
        if now - _last_alert_time < config.alert_cooldown:
            logger.debug("Alert suppressed by cooldown")
            return False

    return True


def _build_payload(result: Any, config: ScanConfig) -> dict:
    """Build a JSON-serialisable alert payload from a scan result."""
    summary = result.summary
    agents = []
    for agent in result.agents_detected:
        d = agent.to_dict()
        agents.append({
            "ip_address": d.get("ip_address", ""),
            "host": d.get("host", ""),
            "port": d.get("port"),
            "agent_type": d.get("agent_type", "unknown"),
            "framework": d.get("framework", "unknown"),
            "confidence_score": d.get("confidence_score", 0),
            "confidence_level": d.get("confidence_level", "low"),
            "status": d.get("status", "unknown"),
            "signal_count": d.get("signal_count", 0),
        })

    return {
        "source": "agentsniff",
        "source_url": "https://agentsniff.org",
        "copyright": "\u00a9 2026 ThirdKey.AI (https://thirdkey.ai)",
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "scan_id": getattr(result, "scan_id", ""),
        "target_network": getattr(result, "target_network", config.target_network),
        "total_agents": summary.get("total_agents", 0),
        "by_confidence": summary.get("by_confidence", {}),
        "duration_seconds": summary.get("duration_seconds"),
        "agents": agents,
    }


async def send_alerts(result: Any, config: ScanConfig) -> list[str]:
    """Dispatch alerts to configured channels. Returns outcome strings."""
    global _last_alert_time

    outcomes: list[str] = []
    payload = _build_payload(result, config)

    if config.webhook_url:
        outcome = await _send_webhook(payload, config)
        outcomes.append(outcome)

    if config.smtp_to:
        outcome = await _send_email(payload, config)
        outcomes.append(outcome)

    if any("ok" in o for o in outcomes):
        _last_alert_time = time.monotonic()

    return outcomes


async def _send_webhook(payload: dict, config: ScanConfig) -> str:
    """POST alert payload to webhook URL."""
    headers = {"Content-Type": "application/json"}
    headers.update(config.webhook_headers)

    try:
        timeout = aiohttp.ClientTimeout(total=15)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.post(
                config.webhook_url, json=payload, headers=headers
            ) as resp:
                if resp.status < 400:
                    logger.info(f"Webhook alert sent ({resp.status})")
                    return f"webhook:ok:{resp.status}"
                body = await resp.text()
                logger.warning(f"Webhook returned {resp.status}: {body[:200]}")
                return f"webhook:failed:{resp.status}"
    except Exception as e:
        logger.error(f"Webhook alert failed: {e}")
        return f"webhook:failed:{e}"


async def _send_email(payload: dict, config: ScanConfig) -> str:
    """Send alert email via SMTP (runs in executor to avoid blocking)."""
    if not config.smtp_host:
        return "email:skipped:no_smtp_host"

    loop = asyncio.get_running_loop()
    try:
        await loop.run_in_executor(None, _smtp_send, payload, config)
        logger.info(f"Email alert sent to {', '.join(config.smtp_to)}")
        return "email:ok"
    except Exception as e:
        logger.error(f"Email alert failed: {e}")
        return f"email:failed:{e}"


def _smtp_send(payload: dict, config: ScanConfig) -> None:
    """Blocking SMTP send — called via run_in_executor."""
    total = payload["total_agents"]
    by_conf = payload.get("by_confidence", {})
    conf_summary = ", ".join(f"{k}: {v}" for k, v in by_conf.items() if v)

    subject = f"[AgentSniff] {total} agent(s) detected on {payload.get('target_network', 'network')}"

    # Plain text body
    lines = [
        "AgentSniff Alert",
        "",
        f"Scan ID:   {payload.get('scan_id', 'N/A')}",
        f"Network:   {payload.get('target_network', 'N/A')}",
        f"Timestamp: {payload.get('timestamp', 'N/A')}",
        f"Duration:  {payload.get('duration_seconds', 'N/A')}s",
        "",
        f"Total Agents: {total}",
        f"By Confidence: {conf_summary}",
        "",
    ]

    for agent in payload.get("agents", []):
        lines.append(
            f"  {agent['ip_address']}"
            f"{':%s' % agent['port'] if agent.get('port') else ''}"
            f"  {agent['agent_type']}"
            f"  {agent['framework']}"
            f"  {agent['confidence_level']} ({agent['confidence_score']:.0%})"
        )

    lines.append("")
    lines.append("--")
    lines.append("AgentSniff (https://agentsniff.org) \u00a9 2026 ThirdKey.AI (https://thirdkey.ai)")

    text_body = "\n".join(lines)

    msg = MIMEMultipart("mixed")
    msg["Subject"] = subject
    msg["From"] = config.smtp_from or config.smtp_user
    msg["To"] = ", ".join(config.smtp_to)
    msg.attach(MIMEText(text_body, "plain"))

    # Attach JSON payload as a file
    json_part = MIMEText(json.dumps(payload, indent=2, default=str), "plain")
    json_part.add_header("Content-Disposition", "attachment", filename="alert-payload.json")
    msg.attach(json_part)

    if config.smtp_use_tls:
        server = smtplib.SMTP(config.smtp_host, config.smtp_port, timeout=15)
        server.starttls()
    else:
        server = smtplib.SMTP(config.smtp_host, config.smtp_port, timeout=15)

    try:
        if config.smtp_user and config.smtp_password:
            server.login(config.smtp_user, config.smtp_password)
        server.sendmail(
            config.smtp_from or config.smtp_user,
            config.smtp_to,
            msg.as_string(),
        )
    finally:
        server.quit()

File: agentsniff-rs/crates/agentsniff/src/notifier.rs
use std::time::Instant;

use lettre::message::{header::ContentType, Mailbox, MultiPart, SinglePart};
use lettre::transport::smtp::authentication::Credentials;
use lettre::{AsyncSmtpTransport, AsyncTransport, Message, Tokio1Executor};

use crate::config::ScanConfig;
use crate::models::{DetectedAgent, ScanResult};

/// Handles webhook and SMTP alerting when agents are detected.
pub struct Notifier {
    config: ScanConfig,
    last_alert: Option<Instant>,
}

impl Notifier {
    pub fn new(config: &ScanConfig) -> Self {
        Self {
            config: config.clone(),
            last_alert: None,
        }
    }

    /// Check if alert conditions are met and send notifications.
    pub async fn check_and_notify(&mut self, result: &ScanResult) -> anyhow::Result<()> {
        if !self.config.alert_enabled {
            return Ok(());
        }

        // Check cooldown
        if let Some(last) = self.last_alert {
            if last.elapsed().as_secs() < self.config.alert_cooldown {
                return Ok(());
            }
        }

        // Check min agents threshold
        let qualifying_agents: Vec<&DetectedAgent> = result
            .agents
            .iter()
            .filter(|a| a.confidence_score >= self.config.alert_min_confidence)
            .collect();

        if qualifying_agents.len() < self.config.alert_min_agents {
            return Ok(());
        }

        let outcomes = send_alerts(result, &self.config).await;
        if outcomes.iter().any(|o| o.contains("ok")) {
            self.last_alert = Some(Instant::now());
        }

        Ok(())
    }
}

/// Build the JSON alert payload from a scan result.
fn build_payload(result: &ScanResult, _config: &ScanConfig) -> serde_json::Value {
    let agents: Vec<serde_json::Value> = result
        .agents
        .iter()
        .map(|a| {
            serde_json::json!({
                "ip_address": a.ip_address.to_string(),
                "host": a.host,
                "port": a.port,
                "agent_type": a.agent_type,
                "framework": a.framework,
                "confidence_score": a.confidence_score,
                "confidence_level": a.confidence_level,
                "status": a.status,
                "signal_count": a.signal_count,
            })
        })
        .collect();

    serde_json::json!({
        "source": "agentsniff",
        "source_url": "https://agentsniff.org",
        "copyright": "\u{00a9} 2026 ThirdKey.AI (https://thirdkey.ai)",
        "timestamp": chrono::Utc::now().to_rfc3339(),
        "scan_id": result.scan_id,
        "target_network": result.target_network.clone(),
        "total_agents": result.agents.len(),
        "duration_seconds": result.duration_seconds(),
        "agents": agents,
    })
}

/// Dispatch alerts to all configured channels. Returns outcome strings.
pub async fn send_alerts(result: &ScanResult, config: &ScanConfig) -> Vec<String> {
    let mut outcomes = Vec::new();
    let payload = build_payload(result, config);

    if !config.webhook_url.is_empty() {
        outcomes.push(send_webhook(&payload, config).await);
    }

    if !config.smtp_to.is_empty() {
        outcomes.push(send_email(&payload, config).await);
    }

    outcomes
}

async fn send_webhook(payload: &serde_json::Value, config: &ScanConfig) -> String {
    let client = reqwest::Client::new();
    let mut req = client
        .post(&config.webhook_url)
        .header("Content-Type", "application/json")
        .json(payload);

    for (key, value) in &config.webhook_headers {
        req = req.header(key.as_str(), value.as_str());
    }

    match req.send().await {
        Ok(resp) => {
            let status = resp.status().as_u16();
            if status < 400 {
                tracing::info!("Webhook alert sent ({})", status);
                format!("webhook:ok:{}", status)
            } else {
                tracing::warn!("Webhook returned {}", status);
                format!("webhook:failed:{}", status)
            }
        }
        Err(e) => {
            tracing::error!("Webhook alert failed: {}", e);
            format!("webhook:failed:{}", e)
        }
    }
}

async fn send_email(payload: &serde_json::Value, config: &ScanConfig) -> String {
    if config.smtp_host.is_empty() {
        return "email:skipped:no_smtp_host".to_string();
    }

    let total = payload["total_agents"].as_u64().unwrap_or(0);
    let network = payload["target_network"].as_str().unwrap_or("network");

    let subject = format!("[AgentSniff] {} agent(s) detected on {}", total, network);

    // Build plain text body
    let mut lines = vec![
        "AgentSniff Alert".to_string(),
        String::new(),
        format!("Scan ID:   {}", payload["scan_id"].as_str().unwrap_or("N/A")),
        format!("Network:   {}", network),
        format!("Timestamp: {}", payload["timestamp"].as_str().unwrap_or("N/A")),
        format!("Duration:  {}s", payload["duration_seconds"]),
        String::new(),
        format!("Total Agents: {}", total),
        String::new(),
    ];

    if let Some(agents) = payload["agents"].as_array() {
        for agent in agents {
            let ip = agent["ip_address"].as_str().unwrap_or("?");
            let port_str = agent["port"]
                .as_u64()
                .map(|p| format!(":{}", p))
                .unwrap_or_default();
            let atype = agent["agent_type"].as_str().unwrap_or("unknown");
            let fw = agent["framework"].as_str().unwrap_or("unknown");
            let level = agent["confidence_level"].as_str().unwrap_or("low");
            let score = agent["confidence_score"].as_f64().unwrap_or(0.0);
            lines.push(format!(
                "  {}{}  {}  {}  {} ({:.0}%)",
                ip,
                port_str,
                atype,
                fw,
                level,
                score * 100.0
            ));
        }
    }

    lines.push(String::new());
    lines.push("--".to_string());
    lines.push(
        "AgentSniff (https://agentsniff.org) \u{00a9} 2026 ThirdKey.AI (https://thirdkey.ai)"
            .to_string(),
    );

    let text_body = lines.join("\n");
    let json_attachment =
        serde_json::to_string_pretty(payload).unwrap_or_else(|_| "{}".to_string());

    let from_addr = if config.smtp_from.is_empty() {
        &config.smtp_user
    } else {
        &config.smtp_from
    };

    let from_mailbox: Mailbox = match from_addr.parse() {
        Ok(m) => m,
        Err(e) => return format!("email:failed:bad_from_address:{}", e),
    };

    let mut to_mailboxes = Vec::new();
    for addr in &config.smtp_to {
        match addr.trim().parse::<Mailbox>() {
            Ok(m) => to_mailboxes.push(m),
            Err(e) => {
                tracing::warn!("Skipping invalid recipient '{}': {}", addr, e);
            }
        }
    }

    if to_mailboxes.is_empty() {
        return "email:failed:no_valid_recipients".to_string();
    }

    let mut builder = Message::builder()
        .from(from_mailbox)
        .subject(subject);

    for mb in &to_mailboxes {
        builder = builder.to(mb.clone());
    }

    let email = match builder.multipart(
        MultiPart::mixed()
            .singlepart(
                SinglePart::builder()
                    .header(ContentType::TEXT_PLAIN)
                    .body(text_body),
            )
            .singlepart(
                SinglePart::builder()
                    .header(ContentType::parse("application/json").unwrap())
                    .header(lettre::message::header::ContentDisposition::attachment(
                        "alert-payload.json",
                    ))
                    .body(json_attachment),
            ),
    ) {
        Ok(msg) => msg,
        Err(e) => return format!("email:failed:build:{}", e),
    };

    // Build SMTP transport
    let transport_result = if config.smtp_use_tls {
        AsyncSmtpTransport::<Tokio1Executor>::starttls_relay(&config.smtp_host)
    } else {
        AsyncSmtpTransport::<Tokio1Executor>::relay(&config.smtp_host)
    };

    let transport = match transport_result {
        Ok(builder) => {
            let mut b = builder.port(config.smtp_port);
            if !config.smtp_user.is_empty() {
                b = b.credentials(Credentials::new(
                    config.smtp_user.clone(),
                    config.smtp_password.clone(),
                ));
            }
            b.build()
        }
        Err(e) => return format!("email:failed:transport:{}", e),
    };

    match transport.send(email).await {
        Ok(_) => {
            tracing::info!(
                "Email alert sent to {}",
                config.smtp_to.join(", ")
            );
            "email:ok".to_string()
        }
        Err(e) => {
            tracing::error!("Email alert failed: {}", e);
            format!("email:failed:{}", e)
        }
    }
}
